File: examples/maven-basic/src/main/java/example/WaveformProcessor.java
package example;

public class WaveformProcessor {

    private int samplingRate;
    private double amplitudeFactor;

    public WaveformProcessor(int samplingRate, double amplitudeFactor) {
        this.samplingRate = samplingRate;
        this.amplitudeFactor = amplitudeFactor;
    }

    public double[] applyAmplification(double[] inputWaveform) {
        double[] outputWaveform = new double[inputWaveform.length];
        for (int sampleIndex = 0; sampleIndex < inputWaveform.length; sampleIndex++) {
            outputWaveform[sampleIndex] = inputWaveform[sampleIndex] * amplitudeFactor;
        }
        return outputWaveform;
    }

    public int countPeaksAboveThreshold(double[] waveform, double threshold) {
        int peakCount = 0;
        for (int idx = 1; idx < waveform.length - 1; idx++) {
            boolean isLocalMaximum = waveform[idx] > waveform[idx - 1] && 
                                     waveform[idx] > waveform[idx + 1];
            if (isLocalMaximum && waveform[idx] > threshold) {
                peakCount++;
            }
        }
        return peakCount;
    }

    public int getSamplingRate() {
        return samplingRate;
    }

    public void setSamplingRate(int samplingRate) {
        this.samplingRate = samplingRate;
    }

    public double getAmplitudeFactor() {
        return amplitudeFactor;
    }

    public void setAmplitudeFactor(double amplitudeFactor) {
        this.amplitudeFactor = amplitudeFactor;
    }
}

File: examples/maven-scala/src/main/scala/example/WaveformProcessor.scala
package example

/**
 * Simple waveform utilities.
 *
 * Case class boilerplate (copy, equals, hashCode, …) will be
 * filtered from coverage via `jmf-rules.txt`.
 */
case class WaveformProcessor(samplingRate: Int, amplitudeFactor: Double) {

  /** Scale every sample by the amplitude factor. */
  def applyAmplification(waveform: Array[Double]): Array[Double] =
    waveform.map(_ * amplitudeFactor)

  /** Count local maxima that exceed the given threshold. */
  def countPeaksAboveThreshold(waveform: Array[Double], threshold: Double): Int =
    waveform.sliding(3).count { window =>
      window(1) > window(0) && window(1) > window(2) && window(1) > threshold
    }
}
